sphinx_automodapi/automodsumm.py

```diff
@@ -236,11 +236,33 @@ def run(self):
         try:
             if len(clsnms) > 0:
                 self.arguments = [' '.join(clsnms)]
-            return InheritanceDiagram.run(self)
+            node = InheritanceDiagram.run(self)[0]
+            node['graph'].__class__ = AutomodGraph  # cast to a custom subclass
+            return [node]
         finally:
             self.arguments = oldargs
 
 
+class AutomodGraph(InheritanceGraph):
+    """
+    Subclass of `sphinx.ext.inheritance_diagram.InheritanceGraph` that fixes the
+    URLs of the inheritance diagram when generated for SVG output.
+    """
+    def generate_dot(self, name, urls={}, env=None,
+                     graph_attrs={}, node_attrs={}, edge_attrs={}):
+        # For SVG output, sphinx.ext.inheritance_diagram.html_visit_inheritance_diagram()
+        # will prepend '../' in front of every URL, which breaks the links given how
+        # sphinx_automodapi embeds inheritance diagrams.  Therefore, we reverse the
+        # prepending before generating the inheritance diagram.
+        if env is not None:
+            graphviz_output_format = env.config.graphviz_output_format.upper()
+            if graphviz_output_format == 'SVG':
+                urls = {k: v[3:] for k, v in urls.items()}
```

Comment:
The suggestion below works for me if I build the sunpy docs and test on a local web server. This way all the hrefs in SVG files only start with `../generated`. If I understand the upstream issue correctly, the URLs for the PNG diagram need to be relative to the HTML file while the URLs inside the SVG diagram need to be relative to the SVG file itself? So [in this line of the `sphinx` codebase](https://github.com/sphinx-doc/sphinx/blob/c40bb7281ec26c833aa7e0f792f063b2d2254ec7/sphinx/ext/inheritance_diagram.py#L404), they should be able to apply the same fix as here (i.e., remove all `"../"` from `child.get('refuri')`)?
```suggestion
                urls = {k: "../" + v.split("../")[-1] for k, v in urls.items()}
```

Reply:
This approach would work fine for the default configuration, but wouldn't be a general solution for all possible configurations of directory structure.

Reply:
Not sure I understand how a different directory structure could make this not work? (Unless people can change the default image directory.) Can we can assume that `child.get('refuri')` will always be rooted at the directory where the docs get generated to (yet prepended with any number of `../`) and that the SVG files will always be in a directory `_images`? 

I.e., docs are generated to a directory `_build/html` with files `_build/html/user_guide/a/b.html` `_build/html/api_docs/x/y/z/module.html` and a diagram at `_build/html/_images/diag.svg`.

Therefore `child.get('refuri')` will be `../../api_docs/x/y/z/module.html#test.class`. So `child.get('refuri').split("../")[-1]` (`api_docs/x/y/z/module.html#test.class`) is always rooted at the directory containing `_images`. So, you can then do `"../" + child.get('refuri').split("../")[-1]` to get the path to `module.html` relative to `diag.svg`.

Also, the bug is definitely in sphinx as I was able to recreate it using only sphinx.

Reply:
My point is that we shouldn't assume that directory structure.  The SVG files might get stored in a sub-sub-folder, or the relative path to the API docs might start with a `'../'`.  Even if Sphinx itself doesn't support such a directory structure, it is so extensible that one just doesn't know.  All of the information to re-path the URL should be available.

If this were a new bug, a hacky solution would probably be fine to get things working in most cases.  Given that this bug is ancient, let's take the extra time to make a proper solution.

Reply:
Yes, you are correct! I was able to produce a directory structure where it didn't work (as the refuri only goes pack to the first directory in common with the page with the inheritance diagram).

Making the following modifications to ext.inheritance_diagram upstream I can get the links to work reliably on a minimal sphinx doc with lots of different directory hierarchies. I think this should be general solution that will work even if other plugins modify the default paths. What do you think?

```
404c404,412
<                 urls[child['reftitle']] = "../" + child.get('refuri')
---
>                 from pathlib import Path
>                 current_directory = Path(current_filename).parent
>                 url = "../" + str(
>                     current_directory /
>                     Path(self.builder.imgpath).parent /
>                     current_directory /
>                     Path(child.get('refuri'))
>                 )
>                 urls[child['reftitle']] = url
```

Reply:
Hmmm. The imgpath part is redundant in that code so maybe this isn't general...

Reply:
Your patch has some issues.  My Sphinx modification is ayshih/sphinx@e711e26dca033dfa5505c32816fb32775bb8c5b2, and #6267 shows that it fixes all SunPy inheritance diagrams.  Can you try it on your test docs?

Reply:
I didn't notice that you updated the sunpy PR with a patched version of sphinx rather than this plugin! 🙃 😆  Yeah, your solution is much more robust across different systems etc and with variations in the location of the images! It works perfectly on my test docs with a big tree of directories with diagrams and API docs at all different levels. I think this bug is now finally solved! Well done! 🥳 There's an old issue upstream https://github.com/sphinx-doc/sphinx/issues/3176 (as well as one opened just 8 hours ago! https://github.com/sphinx-doc/sphinx/issues/10570)

Reply:
sphinx-doc/sphinx#10576 created!